backend/lcfs/utils/spreadsheet_builder.py

```diff
@@ -0,0 +1,143 @@
+            for sheet in self.sheets_data:
+                self._write_sheet_to_excel(writer, sheet)
+
+    def _write_sheet_to_excel(self, writer, sheet):
+        df = pd.DataFrame(sheet["rows"], columns=sheet["columns"])
+        df.to_excel(writer, index=False, sheet_name=sheet["sheet_name"])
+
+        worksheet = writer.sheets[sheet["sheet_name"]]
+
+        # Apply number formatting for numeric columns
+        for col_idx, column in enumerate(df.columns):
+            if df[column].dtype in ['float', 'int']:
+                for row in worksheet.iter_rows(min_row=2, max_row=worksheet.max_row, min_col=col_idx+1, max_col=col_idx+1):
+                    for cell in row:
+                        cell.number_format = '#,##0'
+                        cell.alignment = styles.Alignment(horizontal='left')
+
+        self._apply_excel_styling(writer, sheet)
+
+    def _apply_excel_styling(self, writer, sheet):
+        worksheet = writer.sheets[sheet["sheet_name"]]
+
+        # Apply no border, non-bold font, and left alignment to header cells
+        for cell in worksheet[1]:
+            cell.border = styles.Border(
+                left=styles.Side(style=None),
+                right=styles.Side(style=None),
+                top=styles.Side(style=None),
+                bottom=styles.Side(style=None)
+            )
+            cell.font = styles.Font(bold=False)
+            cell.alignment = styles.Alignment(horizontal='left')
+
+        if sheet["styles"].get("bold_headers"):
+            for cell in worksheet[1]:
+                cell.font = styles.Font(bold=True)
+
+        for i, width in enumerate(sheet["styles"].get("column_widths", []), start=1):
+            worksheet.column_dimensions[chr(64 + i)].width = width
+
+    def _write_xls(self, output):
+        book = xlwt.Workbook()
+
+        for sheet_data in self.sheets_data:
+            self._write_sheet_to_xls(book, sheet_data)
+        book.save(output)
+
+    def _write_sheet_to_xls(self, book, sheet_data):
+        sheet = book.add_sheet(sheet_data["sheet_name"])
+
+        # Styles for headers
+        bold_style = xlwt.XFStyle()
+        font = xlwt.Font()
+        font.bold = True
+        bold_style.font = font
+        default_style = xlwt.XFStyle()
+
+        # Style for left-aligned numbers
+        left_aligned_num_style = xlwt.XFStyle()
+        left_aligned_num_style.num_format_str = '#,##0'
+        alignment = xlwt.Alignment()
+        alignment.horz = xlwt.Alignment.HORZ_LEFT
+        left_aligned_num_style.alignment = alignment
+
+        # Apply bold style for headers if required
+        header_style = bold_style if sheet_data["styles"].get("bold_headers") else default_style
+
+        for col_index, column in enumerate(sheet_data["columns"]):
+            sheet.write(0, col_index, column, header_style)
+            if "column_widths" in sheet_data["styles"] and len(sheet_data["styles"]["column_widths"]) > col_index:
```

Comment:
I'm thinking we try and make the column_widths dynamic based on the widest cell in each column. Could you try something like the following and let me know what you think.

```python
from openpyxl.utils import get_column_letter

def _auto_adjust_column_width_xlsx(self, writer, sheet_data):
    worksheet = writer.sheets[sheet_data["sheet_name"]]
    for column in worksheet.columns:
        max_length = 0
        column_letter = get_column_letter(column[0].column)

        for cell in column:
            try:
                if len(str(cell.value)) > max_length:
                    max_length = len(cell.value)
            except:
                pass

        adjusted_width = (max_length + 2)  # Adding 2 for a little extra space
        worksheet.column_dimensions[column_letter].width = adjusted_width

Reply:
That's a great idea! I was thinking about it, but wasn't sure if I was allowed to do so. Now, it's better for both developers and users. Thanks for bringing it up.